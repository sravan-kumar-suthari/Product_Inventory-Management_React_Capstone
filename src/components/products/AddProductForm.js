import React from "react";
import { Formik } from "formik";
import * as Yup from "yup";
import { Button } from 'react-bootstrap';
import { Prompt, withRouter } from 'react-router';
import {connect} from "react-redux"
import {addProduct} from "../redux/actions/productActions"
//import {withRouter} from "react-router-dom"

const validationSchema = Yup.object({
  productName: Yup.string().required("Product Name is required"),
  quantity: Yup.number().required("Quantity is required"),
  price: Yup.number().required("Price is required"),
  manufacturer:Yup.string().required("Manufacturer details required"),
  description:Yup.string().required("Descriptopn Needed")
});

class AddProductForm extends React.Component {
  constructor(props) {
  super(props);
    this.state={
        isSubmitted : false
    }
  }

componentDidMount(){
  this.setState({isSubmitted:false});
}

    render() {
      const errorStyle={"color":"red","fontWeight":"bold"}
    
    
    return (
        
      <Formik
        initialValues={{ productName: "", quantity: "", price: "",manufacturer:"",description:"" }}
        validationSchema={validationSchema}
        validateOnChange={false}
        validateOnBlur={false}


        onSubmit={(value) => {
        this.setState({isSubmitted : true   })
        const values={...value, views : 0}
        alert("Product Added Successfully")
        this.props.addProduct(values);
        this.props.history.push("/");
        }
    }
      >

        

        {({ handleSubmit, handleChange, values, errors,touched }) => (
          <form onSubmit={handleSubmit} style={{"margin": "0 auto","width":"30%"}}>

        
            {<Prompt when={
            (values.productName !==""|| values.quantity !=="" || values.price !=="" || values.manufacturer !=="" || values.description !=="") && (this.state.isSubmitted===false) 
            } message="You have unsaved changes, are you sure you want to leave?"/>}

          <h4 style={{"display":"flex","justifyContent":"center","fontWeight":"bold"}}>Add Product</h4>

            <div className="form-group">
            <label htmlFor="name" >Name </label>
            <input
              type="text"
              onChange={handleChange}
              value={values.productName}
              placeholder="Enter Product Name"
              name="productName"
              id="productName"
              className="form-control"
            />
            <span style={errorStyle}>{errors.productName}</span>
            </div>

            <div className="form-group">
            <label htmlFor="quantity" >Quantity </label>
            <input
              type="number"
              onChange={handleChange}
              value={values.quantity}
              placeholder="Enter Quantity"
              name="quantity"
              id="quantity"
              className="form-control"
            />
            <span style={errorStyle}>{errors.quantity}</span>
            </div>

            <div className="form-group">
            <label htmlFor="price" >Price  </label>
            <input
              type="number"
              step="1"
              onChange={handleChange}
              value={values.price}
              placeholder="Enter Price"
              name="price"
              id="price"
              className="form-control"
            />
            <span style={errorStyle}>{errors.price}</span>
            </div>

            <div className="form-group">
            <label htmlFor="description" >Description </label>
            <input
              type="text"
              onChange={handleChange}
              value={values.description}
              placeholder="Enter description"
              name="description"
              id="description"
              className="form-control"
            />
            <span style={errorStyle}>{errors.description}</span>
            </div>

            <div className="form-group">
            <label htmlFor="manufacturer" >Manufacturer </label>
            <input
              type="text"
              onChange={handleChange}
              value={values.manufacturer}
              placeholder="Enter manufacturer"
              name="manufacturer"
              id="manufacturer"
              className="form-control"
            />
            <span style={errorStyle}>{errors.manufacturer}</span>
            </div>

            
            <div style={{"display":"flex","justifyContent":"center"}}><Button type="submit" style={{"fontWeight":"bold"}}>Submit</Button></div>
          </form>
        )}
      </Formik>
    );
  }
}

const mapDispatchToProps=(dispatch)=>{
    return{
        addProduct : product=>dispatch(addProduct(product))
    }
}

export default connect(null,mapDispatchToProps)(withRouter(AddProductForm));