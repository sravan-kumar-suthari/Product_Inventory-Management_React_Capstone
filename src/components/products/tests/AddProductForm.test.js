import React from "react";
//import moxios from "moxios";
import AddProductForm from "../AddProductForm";
import {mount} from "enzyme";
import store from "../../redux/store/configureStore";
import {Provider} from "react-redux"
import {BrowserRouter} from "react-router-dom"
//


describe("AddProductForm snapshot",()=>{
    let mountWrapper;

    beforeEach(()=>{
        mountWrapper=mount(<Provider store={store}><BrowserRouter><AddProductForm /></BrowserRouter> </Provider>)
    })

    afterEach(()=>{
        mountWrapper.unmount();
    })

    test("AddProductForm snapshot renders correctly",()=>{
        expect(mountWrapper).toMatchSnapshot();
        
    })
})

describe("AddProductForm rendering of elements",()=>{
    let mountWrapper;
    beforeEach(()=>{
        mountWrapper=mount(<Provider store={store}><BrowserRouter><AddProductForm /></BrowserRouter> </Provider>)
    })  
    afterEach(()=>{
    mountWrapper.unmount();
})

it("should display 'Add Product' as heading",()=>{
    expect(mountWrapper.find('h4').render().text()).toEqual('Add Product')
})

it("should contain 5 input Elements",()=>{
    expect(mountWrapper.find('input').length).toBe(5);
})

it("should contain one submit button",()=>{
    expect(mountWrapper.find('Button').length).toBe(1)
})

})

