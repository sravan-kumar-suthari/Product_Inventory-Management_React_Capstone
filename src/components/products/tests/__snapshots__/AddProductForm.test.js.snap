// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddProductForm snapshot AddProductForm snapshot renders correctly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(withRouter(AddProductForm))>
        <withRouter(AddProductForm)
          addProduct={[Function]}
        >
          <AddProductForm
            addProduct={[Function]}
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "createHref": [Function],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            location={
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              }
            }
            match={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
          >
            <Formik
              initialValues={
                Object {
                  "description": "",
                  "manufacturer": "",
                  "price": "",
                  "productName": "",
                  "quantity": "",
                }
              }
              onSubmit={[Function]}
              validateOnBlur={false}
              validateOnChange={false}
              validationSchema={
                ObjectSchema {
                  "_blacklist": ReferenceSet {
                    "list": Set {},
                    "refs": Map {},
                  },
                  "_excludedEdges": Array [],
                  "_mutate": undefined,
                  "_nodes": Array [
                    "description",
                    "manufacturer",
                    "price",
                    "quantity",
                    "productName",
                  ],
                  "_sortErrors": [Function],
                  "_typeError": [Function],
                  "_whitelist": ReferenceSet {
                    "list": Set {},
                    "refs": Map {},
                  },
                  "conditions": Array [],
                  "deps": Array [],
                  "exclusiveTests": Object {},
                  "fields": Object {
                    "description": StringSchema {
                      "_blacklist": ReferenceSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "_blacklistError": undefined,
                      "_mutate": undefined,
                      "_typeError": [Function],
                      "_whitelist": ReferenceSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "_whitelistError": undefined,
                      "conditions": Array [],
                      "deps": Array [],
                      "exclusiveTests": Object {
                        "required": true,
                      },
                      "spec": Object {
                        "abortEarly": true,
                        "label": undefined,
                        "meta": undefined,
                        "nullable": false,
                        "presence": "required",
                        "recursive": true,
                        "strict": false,
                        "strip": false,
                      },
                      "tests": Array [
                        [Function],
                      ],
                      "transforms": Array [
                        [Function],
                      ],
                      "type": "string",
                    },
                    "manufacturer": StringSchema {
                      "_blacklist": ReferenceSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "_blacklistError": undefined,
                      "_mutate": undefined,
                      "_typeError": [Function],
                      "_whitelist": ReferenceSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "_whitelistError": undefined,
                      "conditions": Array [],
                      "deps": Array [],
                      "exclusiveTests": Object {
                        "required": true,
                      },
                      "spec": Object {
                        "abortEarly": true,
                        "label": undefined,
                        "meta": undefined,
                        "nullable": false,
                        "presence": "required",
                        "recursive": true,
                        "strict": false,
                        "strip": false,
                      },
                      "tests": Array [
                        [Function],
                      ],
                      "transforms": Array [
                        [Function],
                      ],
                      "type": "string",
                    },
                    "price": NumberSchema {
                      "_blacklist": ReferenceSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "_blacklistError": undefined,
                      "_mutate": undefined,
                      "_typeError": [Function],
                      "_whitelist": ReferenceSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "_whitelistError": undefined,
                      "conditions": Array [],
                      "deps": Array [],
                      "exclusiveTests": Object {
                        "required": true,
                      },
                      "spec": Object {
                        "abortEarly": true,
                        "label": undefined,
                        "meta": undefined,
                        "nullable": false,
                        "presence": "required",
                        "recursive": true,
                        "strict": false,
                        "strip": false,
                      },
                      "tests": Array [
                        [Function],
                      ],
                      "transforms": Array [
                        [Function],
                      ],
                      "type": "number",
                    },
                    "productName": StringSchema {
                      "_blacklist": ReferenceSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "_blacklistError": undefined,
                      "_mutate": undefined,
                      "_typeError": [Function],
                      "_whitelist": ReferenceSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "_whitelistError": undefined,
                      "conditions": Array [],
                      "deps": Array [],
                      "exclusiveTests": Object {
                        "required": true,
                      },
                      "spec": Object {
                        "abortEarly": true,
                        "label": undefined,
                        "meta": undefined,
                        "nullable": false,
                        "presence": "required",
                        "recursive": true,
                        "strict": false,
                        "strip": false,
                      },
                      "tests": Array [
                        [Function],
                      ],
                      "transforms": Array [
                        [Function],
                      ],
                      "type": "string",
                    },
                    "quantity": NumberSchema {
                      "_blacklist": ReferenceSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "_blacklistError": undefined,
                      "_mutate": undefined,
                      "_typeError": [Function],
                      "_whitelist": ReferenceSet {
                        "list": Set {},
                        "refs": Map {},
                      },
                      "_whitelistError": undefined,
                      "conditions": Array [],
                      "deps": Array [],
                      "exclusiveTests": Object {
                        "required": true,
                      },
                      "spec": Object {
                        "abortEarly": true,
                        "label": undefined,
                        "meta": undefined,
                        "nullable": false,
                        "presence": "required",
                        "recursive": true,
                        "strict": false,
                        "strip": false,
                      },
                      "tests": Array [
                        [Function],
                      ],
                      "transforms": Array [
                        [Function],
                      ],
                      "type": "number",
                    },
                  },
                  "spec": Object {
                    "abortEarly": true,
                    "label": undefined,
                    "meta": undefined,
                    "nullable": false,
                    "presence": "optional",
                    "recursive": true,
                    "strict": false,
                    "strip": false,
                  },
                  "tests": Array [],
                  "transforms": Array [
                    [Function],
                  ],
                  "type": "object",
                }
              }
            >
              <form
                onSubmit={[Function]}
                style={
                  Object {
                    "margin": "0 auto",
                    "width": "30%",
                  }
                }
              >
                <Prompt
                  message="You have unsaved changes, are you sure you want to leave?"
                  when={false}
                />
                <h4
                  style={
                    Object {
                      "display": "flex",
                      "fontWeight": "bold",
                      "justifyContent": "center",
                    }
                  }
                >
                  Add Product
                </h4>
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="name"
                  >
                    Name 
                  </label>
                  <input
                    className="form-control"
                    id="productName"
                    name="productName"
                    onChange={[Function]}
                    placeholder="Enter Product Name"
                    type="text"
                    value=""
                  />
                  <span
                    style={
                      Object {
                        "color": "red",
                        "fontWeight": "bold",
                      }
                    }
                  />
                </div>
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="quantity"
                  >
                    Quantity 
                  </label>
                  <input
                    className="form-control"
                    id="quantity"
                    name="quantity"
                    onChange={[Function]}
                    placeholder="Enter Quantity"
                    type="number"
                    value=""
                  />
                  <span
                    style={
                      Object {
                        "color": "red",
                        "fontWeight": "bold",
                      }
                    }
                  />
                </div>
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="price"
                  >
                    Price  
                  </label>
                  <input
                    className="form-control"
                    id="price"
                    name="price"
                    onChange={[Function]}
                    placeholder="Enter Price"
                    step="1"
                    type="number"
                    value=""
                  />
                  <span
                    style={
                      Object {
                        "color": "red",
                        "fontWeight": "bold",
                      }
                    }
                  />
                </div>
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="description"
                  >
                    Description 
                  </label>
                  <input
                    className="form-control"
                    id="description"
                    name="description"
                    onChange={[Function]}
                    placeholder="Enter description"
                    type="text"
                    value=""
                  />
                  <span
                    style={
                      Object {
                        "color": "red",
                        "fontWeight": "bold",
                      }
                    }
                  />
                </div>
                <div
                  className="form-group"
                >
                  <label
                    htmlFor="manufacturer"
                  >
                    Manufacturer 
                  </label>
                  <input
                    className="form-control"
                    id="manufacturer"
                    name="manufacturer"
                    onChange={[Function]}
                    placeholder="Enter manufacturer"
                    type="text"
                    value=""
                  />
                  <span
                    style={
                      Object {
                        "color": "red",
                        "fontWeight": "bold",
                      }
                    }
                  />
                </div>
                <div
                  style={
                    Object {
                      "display": "flex",
                      "justifyContent": "center",
                    }
                  }
                >
                  <Button
                    active={false}
                    disabled={false}
                    style={
                      Object {
                        "fontWeight": "bold",
                      }
                    }
                    type="submit"
                    variant="primary"
                  >
                    <button
                      className="btn btn-primary"
                      disabled={false}
                      style={
                        Object {
                          "fontWeight": "bold",
                        }
                      }
                      type="submit"
                    >
                      Submit
                    </button>
                  </Button>
                </div>
              </form>
            </Formik>
          </AddProductForm>
        </withRouter(AddProductForm)>
      </Connect(withRouter(AddProductForm))>
    </Router>
  </BrowserRouter>
   
</Provider>
`;
